app/scripts/services/serviceInstances.js

```diff
@@ -2,65 +2,129 @@
                                                DataService,
+                                               Logger,
                                                NotificationsService) {
+    var getBindingsIfNecessary = function(apiObject, bindings) {
+      var deferred = $q.defer();
 
-    function deprovision(apiObject) {
-      var getErrorDetails = $filter('getErrorDetails');
-      var modalScope = {
-        alerts: {
-          deprovision: {
-            type: 'error',
-            message: 'Service \'' + apiObject.spec.serviceClassName + '\' will be deleted and no longer available.'
+      if (angular.isDefined(bindings)) {
+        return $q.when(bindings);
+      } else {
+        var context = {namespace: apiObject.metadata.namespace};
+        var resource = {
+          group: 'servicecatalog.k8s.io',
+          resource: 'serviceinstancecredentials'
+        };
+
+        DataService.list(resource, context, function (serviceBindings) {
+          bindings = serviceBindings.by('metadata.name');
+          bindings = BindingService.getBindingsForResource(bindings, apiObject);
+          deferred.resolve(bindings);
+        });
+      }
+
+      return deferred.promise;
+    };
+
+    var deprovisionInstance = function(apiObject) {
+      var context = {namespace: apiObject.metadata.namespace};
+      var resource = {
+        group: 'servicecatalog.k8s.io',
+        resource: 'serviceinstances'
+      };
+
+      NotificationsService.hideNotification("deprovision-service-error");
+
+      // TODO - remove once this is resolved https://github.com/kubernetes-incubator/service-catalog/issues/942
+      var opts = {
+        propagationPolicy: null
+      };
+
+      return DataService.delete(resource, apiObject.metadata.name, context, opts).then(function() {
+        NotificationsService.addNotification({
+          type: "success",
+          message: "Provisioned service '" + apiObject.metadata.name + "' was marked for deletion."
+        });
+      }, function(err) {
+        NotificationsService.addNotification({
+          id: "deprovision-service-error",
+          type: "error",
+          message: "An error occurred while deleting provisioned service " + apiObject.metadata.name + ".",
+          details: $filter('getErrorDetails')(err)
+        });
+        Logger("An error occurred while deleting provisioned service " + apiObject.metadata.name + ".", err);
+      });
+    };
+
+    var deprovisionBindings = function(apiObject, bindings) {
+      if (!CatalogService.SERVICE_CATALOG_ENABLED) {
+        return;
+      }
+
+      var context = {namespace: apiObject.metadata.namespace};
+      var resource = {
+        group: 'servicecatalog.k8s.io',
+        resource: 'serviceinstancecredentials'
+      };
+
+      getBindingsIfNecessary(apiObject, bindings).then(function(serviceBindings) {
+        _.each(serviceBindings, function (binding) {
+          if (!binding.metadata.deletionTimestamp) {
```

Comment:
This is a style nit, but I personally find early returns easier to reason about.

```js
if (binding.metadata.deletionTimestamp) {
  // The binding is marked for deletion.
  return;
}

// <code to delete the binding>
```

Reply:
fixed